README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->
## Introduction

The package is inspired by the window presentation form design toolbar. It will open the dropdown menu on hover, you can also use it where you want your dropdown to open when the user hovers their mouse, It will show and hide the dropdown menu when mouse enters or leave the widget.

## Features

1. Show/hide dropdown menu on hover.
2. Customize your design with respect to you theme

Example

<!-- https://user-images.githubusercontent.com/26992726/169668343-0b7332dc-04fc-4b03-8a48-0046ae8f126f.mp4 -->

![example](https://user-images.githubusercontent.com/26992726/169691894-197bbe96-e519-4c06-a595-41a912dd6386.gif)


![image](https://user-images.githubusercontent.com/26992726/169668224-8175ae0e-466e-419d-8fac-f27b496d7b24.png)

## Getting started

**1.** Install **window_style_dropdown_menu**:
```bash
flutter pub get window_style_dropdown_menu
```

**2.** Import **window_style_dropdown_menu** in your main.dart file:
```bash
import 'package:window_style_dropdown_menu/window_style_dropdown_menu.dart';
```

<!-- TODO: List prerequisites and provide or point to information on how to
start using the package. -->

## Usage

```dart
import 'package:flutter/material.dart';
import 'package:window_style_dropdown_menu/window_style_dropdown_menu.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Window Style Dropdown Menu',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Window Style Dropdown Menu'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          Container(
            width: double.infinity,
            decoration: const BoxDecoration(
                border:
                    Border(bottom: BorderSide(color: Colors.black, width: 1))),
            height: 50,
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'File',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + O',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Open',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + N',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'New',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
                const VerticalDivider(
                  width: 2,
                  thickness: 1,
                  color: Colors.white,
                ),
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'Save',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + S',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Save',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + K + S',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Save All',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
                const VerticalDivider(
                  width: 2,
                  thickness: 1,
                  color: Colors.white,
                ),
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'Settings',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + E',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Extensions',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + E',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Editor Settings',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
                const VerticalDivider(
                  width: 2,
                  thickness: 1,
                  color: Colors.white,
                ),
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'Others',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + T',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Tools',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + Shift + P',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Command Palette',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
              ],
            ),
          ),
          const Expanded(
              child: Center(
            child: Text('Window Style Dropdown Menu'),
          ))
        ],
      ),
    );
  }
}
```

## Additional information
<!-- 
TODO: Tell users more about the package: where to find more information, how to 
contribute to the package, how to file issues, what response they can expect 
from the package authors, and more. -->


CHANGELOG.md
## 0.0.1

* Initial Release

## 1.0.0

* First Release

test/window_style_dropdown_menu_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:window_style_dropdown_menu/window_style_dropdown_menu.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


lib/window_style_dropdown_menu.dart
library window_style_dropdown_menu;

import 'package:flutter/material.dart';

/// Its is an statefull widget
class WindowStyleDropdownMenu extends StatefulWidget {
  /// Title for the dropdown
  ///
  /// Only accepts string value
  String buttonTitle;

  /// To Style the [buttonTitle] property
  ///
  /// Only accepts [TextStyle]
  TextStyle? buttonTitleStyle;

  /// To customise the dropdown menu with
  ///
  /// by default width is 200.0
  /// Only acceps [double] value
  double? dropdownWidth;

  /// Use to change the dropdown background color
  ///
  /// Only accepts [Color] class
  /// It's an optional field
  /// By default color is [grey]
  Color? dropdownBackgroundColor;

  /// Dropdown items to be shown in the menu
  ///
  /// expects list of [ListTile] class
  List<ListTile> dropdownItems;

  WindowStyleDropdownMenu(
      {Key? key,
      required this.buttonTitle,
      required this.dropdownItems,
      this.buttonTitleStyle,
      this.dropdownWidth,
      this.dropdownBackgroundColor})
      : super(key: key);

  @override
  State<WindowStyleDropdownMenu> createState() =>
      _WindowStyleDropdownMenuState();
}

class _WindowStyleDropdownMenuState extends State<WindowStyleDropdownMenu> {
  OverlayEntry? _overlayEntry;
  FocusNode textFocusNode = FocusNode();
  bool showOverlay = false;

  @override
  void initState() {
    super.initState();
    textFocusNode.addListener(() {
      if (textFocusNode.hasFocus) {
        _overlayEntry = _createOverlayEntry();
        Overlay.of(context)?.insert(_overlayEntry!);
        // _showOverlay(context, 0);
      } else {
        removeOverlay();
      }
    });
  }

  void removeOverlay() {
    _overlayEntry!.remove();
  }

  @override
  Widget build(BuildContext context) {
    return TextButton(
      focusNode: textFocusNode,
      onHover: (val) {
        if (val) {
          textFocusNode.requestFocus();
          showOverlay = true;
        }
      },
      onPressed: () {},
      child: Text(
        widget.buttonTitle,
        style: widget.buttonTitleStyle ?? const TextStyle(color: Colors.white),
      ),
    );
  }

  OverlayEntry _createOverlayEntry() {
    RenderBox renderBox = context.findRenderObject() as RenderBox;
    var size = renderBox.size;
    var offset = renderBox.localToGlobal(Offset.zero);

    return OverlayEntry(
        maintainState: true,
        builder: (context) => Positioned(
              left: offset.dx,
              top: offset.dy + size.height,
              width: widget.dropdownWidth ?? 200,
              child: TextButton(
                onPressed: () {},
                onHover: (val) {
                  if (val && showOverlay) {
                    textFocusNode.requestFocus();
                  } else {
                    textFocusNode.unfocus();
                  }
                },
                child: Material(
                  color: widget.dropdownBackgroundColor ?? Colors.grey,
                  elevation: 4.0,
                  child: ListView(
                      padding: EdgeInsets.zero,
                      shrinkWrap: true,
                      children: widget.dropdownItems),
                ),
              ),
            ));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:window_style_dropdown_menu/window_style_dropdown_menu.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Window Style Dropdown Menu',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Window Style Dropdown Menu'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          Container(
            width: double.infinity,
            decoration: const BoxDecoration(
                border:
                    Border(bottom: BorderSide(color: Colors.black, width: 1))),
            height: 50,
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'File',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + O',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Open',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + N',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'New',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
                const VerticalDivider(
                  width: 2,
                  thickness: 1,
                  color: Colors.white,
                ),
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'Save',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + S',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Save',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + K + S',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Save All',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
                const VerticalDivider(
                  width: 2,
                  thickness: 1,
                  color: Colors.white,
                ),
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'Settings',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + E',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Extensions',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + E',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Editor Settings',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
                const VerticalDivider(
                  width: 2,
                  thickness: 1,
                  color: Colors.white,
                ),
                WindowStyleDropdownMenu(
                  buttonTitleStyle: const TextStyle(color: Colors.black),
                  dropdownWidth: 300,
                  buttonTitle: 'Others',
                  dropdownItems: const [
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + T',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Tools',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                    ListTile(
                      mouseCursor: SystemMouseCursors.click,
                      trailing: Text('Ctrl + Shift + P',
                          style: TextStyle(color: Colors.white)),
                      title: Text(
                        'Command Palette',
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
              ],
            ),
          ),
          const Expanded(
              child: Center(
            child: Text('Window Style Dropdown Menu'),
          ))
        ],
      ),
    );
  }
}
